### PR TITLE
Support AirMentor 2S

### LINKED ISSUE #1100: [New Sensor]: Air Mentor 2S
### Sensor Description

Air Mentor 2S

### Additional information

https://www.air-mentor.com/web2017/product/air_mentor_8099af

### BLE advertisements

```shell
INFO (Thread-2) [custom_components.ble_monitor.ble_parser] Unknown advertisement received for mac: xx:xx:xx:xx:xx:xxservice data: []manufacturer specific data: [b'\x0f\xff\x11a\x02\xb6\x00\x07\x00\n\x1b\x94\x00\x12H\x01']local name: 8099-AFUUID16: None,UUID128: None
INFO (Thread-2) [custom_components.ble_monitor.ble_parser] Unknown advertisement received for mac: xx:xx:xx:xx:xx:xxservice data: []manufacturer specific data: [b'\x0f\xff\x11a\x02\xb6\x00\x07\x00\t\x1b\x94\x00\x12M\x01']local name: 8099-AFUUID16: None,UUID128: None
INFO (Thread-2) [custom_components.ble_monitor.ble_parser] Unknown advertisement received for mac: xx:xx:xx:xx:xx:xxservice data: []manufacturer specific data: [b'\x0f\xff\x12Q\x03\xb9\x19\xadE(\x00\x90\x05U\x00\x02']local name: 8099-AFUUID16: None,UUID128: None
INFO (Thread-2) [custom_components.ble_monitor.ble_parser] Unknown advertisement received for mac: xx:xx:xx:xx:xx:xxservice data: []manufacturer specific data: [b'\x0f\xff\x11a\x02\xb6\x00\x07\x00\t\x1b\x94\x00\x12N\x01']local name: 8099-AFUUID16: None,UUID128: None
INFO (Thread-2) [custom_components.ble_monitor.ble_parser] Unknown advertisement received for mac: xx:xx:xx:xx:xx:xxservice data: []manufacturer specific data: [b'\x0f\xff\x12Q\x03\xd5\x19\xabE(\x00\x93\x05U\x00\x02']local name: 8099-AFUUID16: None,UUID128: None
INFO (Thread-2) [custom_components.ble_monitor.ble_parser] Unknown advertisement received for mac: xx:xx:xx:xx:xx:xxservice data: []manufacturer specific data: [b'\x0f\xff\x11a\x02\xb6\x00\x06\x00\x08\x1b\x94\x00\x12O\x01']local name: 8099-AFUUID16: None,UUID128: None
INFO (Thread-2) [custom_components.ble_monitor.ble_parser] Unknown advertisement received for mac: xx:xx:xx:xx:xx:xxservice data: []manufacturer specific data: [b'\x0f\xff\x12Q\x03\xc4\x19\xabE(\x00\x91\x05U\x00\x02']local name: 8099-AFUUID16: None,UUID128: None
INFO (Thread-2) [custom_components.ble_monitor.ble_parser] Unknown advertisement received for mac: xx:xx:xx:xx:xx:xxservice data: []manufacturer specific data: [b'\x0f\xff\x12Q\x03\xc4\x19\xabE(\x00\x91\x05U\x00\x02']local name: 8099-AFUUID16: None,UUID128: None
INFO (Thread-2) [custom_components.ble_monitor.ble_parser] Unknown advertisement received for mac: xx:xx:xx:xx:xx:xxservice data: []manufacturer specific data: [b'\x0f\xff\x11a\x02\xb6\x00\x06\x00\x08\x1b\x94\x00\x12P\x01']local name: 8099-AFUUID16: None,UUID128: None
INFO (Thread-2) [custom_components.ble_monitor.ble_parser] Unknown advertisement received for mac: xx:xx:xx:xx:xx:xxservice data: []manufacturer specific data: [b'\x0f\xff\x12Q\x03\xca\x19\xabE(\x00\x92\x05U\x00\x02']local name: 8099-AFUUID16: None,UUID128: None
INFO (Thread-2) [custom_components.ble_monitor.ble_parser] Unknown advertisement received for mac: xx:xx:xx:xx:xx:xxservice data: []manufacturer specific data: [b'\x0f\xff\x12Q\x03\xca\x19\xabE(\x00\x92\x05U\x00\x02']local name: 8099-AFUUID16: None,UUID128: None
INFO (Thread-2) [custom_components.ble_monitor.ble_parser] Unknown advertisement received for mac: xx:xx:xx:xx:xx:xxservice data: []manufacturer specific data: [b'\x0f\xff\x11a\x02\xb6\x00\x06\x00\x08\x1b\x94\x00\x12Q\x01']local name: 8099-AFUUID16: None,UUID128: None
INFO (Thread-2) [custom_components.ble_monitor.ble_parser] Unknown advertisement received for mac: xx:xx:xx:xx:xx:xxservice data: []manufacturer specific data: [b'\x0f\xff\x12Q\x03\xdf\x19\xadE(\x00\x94\x05U\x00\x02']local name: 8099-AFUUID16: None,UUID128: None
INFO (Thread-2) [custom_components.ble_monitor.ble_parser] Unknown advertisement received for mac: xx:xx:xx:xx:xx:xxservice data: []manufacturer specific data: [b'\x0f\xff\x11a\x02\xb6\x00\x04\x00\x05\x1b\x94\x00\x12`\x01']local name: 8099-AFUUID16: None,UUID128: None
INFO (Thread-2) [custom_components.ble_monitor.ble_parser] Unknown advertisement received for mac: xx:xx:xx:xx:xx:xxservice data: []manufacturer specific data: [b'\x0f\xff"Q\x03\xc7\x19\xaeE(\x00\x92\x05U\x00\x02']local name: 8099-AFUUID16: None,UUID128: None
INFO (Thread-2) [custom_components.ble_monitor.ble_parser] Unknown advertisement received for mac: xx:xx:xx:xx:xx:xxservice data: []manufacturer specific data: [b'\x0f\xff!a\x02\xb6\x00\x04\x00\x05\x1b\x94\x00\x12a\x01']local name: 8099-AFUUID16: None,UUID128: None
INFO (Thread-2) [custom_components.ble_monitor.ble_parser] Unknown advertisement received for mac: xx:xx:xx:xx:xx:xxservice data: []manufacturer specific data: [b'\x0f\xff"Q\x03\xc5\x19\xaeE(\x00\x92\x05U\x00\x02']local name: 8099-AFUUID16: None,UUID128: None
INFO (Thread-2) [custom_components.ble_monitor.ble_parser] Unknown advertisement received for mac: xx:xx:xx:xx:xx:xxservice data: []manufacturer specific data: [b'\x0f\xff!a\x02\xb7\x00\x04\x00\x05\x1b\x94\x00\x12b\x01']local name: 8099-AFUUID16: None,UUID128: None
INFO (Thread-2) [custom_components.ble_monitor.ble_parser] Unknown advertisement received for mac: xx:xx:xx:xx:xx:xxservice data: []manufacturer specific data: [b'\x0f\xff"Q\x03\xc3\x19\xaeE(\x00\x91\x05U\x00\x02']local name: 8099-AFUUID16: None,UUID128: None
INFO (Thread-2) [custom_components.ble_monitor.ble_parser] Unknown advertisement received for mac: xx:xx:xx:xx:xx:xxservice data: []manufacturer specific data: [b'\x0f\xff!a\x02\xb6\x00\x04\x00\x05\x1b\x94\x00\x12c\x01']local name: 8099-AFUUID16: None,UUID128: None
INFO (Thread-2) [custom_components.ble_monitor.ble_parser] Unknown advertisement received for mac: xx:xx:xx:xx:xx:xxservice data: []manufacturer specific data: [b'\x0f\xff"Q\x03\xda\x19\xadE(\x00\x94\x05U\x00\x02']local name: 8099-AFUUID16: None,UUID128: None
INFO (Thread-2) [custom_components.ble_monitor.ble_parser] Unknown advertisement received for mac: xx:xx:xx:xx:xx:xxservice data: []manufacturer specific data: [b'\x0f\xff!a\x02\xb6\x00\x03\x00\x04\x1b\x94\x00\x12d\x01']local name: 8099-AFUUID16: None,UUID128: None
INFO (Thread-2) [custom_components.ble_monitor.ble_parser] Unknown advertisement received for mac: xx:xx:xx:xx:xx:xxservice data: []manufacturer specific data: [b'\x0f\xff"Q\x03\xcb\x19\xadE(\x00\x92\x05U\x00\x02']local name: 8099-AFUUID16: None,UUID128: None
INFO (Thread-2) [custom_components.ble_monitor.ble_parser] Unknown advertisement received for mac: xx:xx:xx:xx:xx:xxservice data: []manufacturer specific data: [b'\x0f\xff!a\x02\xb7\x00\x03\x00\x04\x1b\x94\x00\x12e\x01']local name: 8099-AFUUID16: None,UUID128: None
INFO (Thread-2) [custom_components.ble_monitor.ble_parser] Unknown advertisement received for mac: xx:xx:xx:xx:xx:xxservice data: []manufacturer specific data: [b'\x0f\xff"Q\x03\xdc\x19\xadE&\x00\x94\x05U\x00\x02']local name: 8099-AFUUID16: None,UUID128: None
INFO (Thread-2) [custom_components.ble_monitor.ble_parser] Unknown advertisement received for mac: xx:xx:xx:xx:xx:xxservice data: []manufacturer specific data: [b'\x0f\xff!a\x02\xb7\x00\x03\x00\x04\x1b\x94\x00\x12f\x01']local name: 8099-AFUUID16: None,UUID128: None
INFO (Thread-2) [custom_components.ble_monitor.ble_parser] Unknown advertisement received for mac: xx:xx:xx:xx:xx:xxservice data: []manufacturer specific data: [b'\x0f\xff"Q\x03\xc1\x19\xb0E(\x00\x91\x05U\x00\x02']local name: 8099-AFUUID16: None,UUID128: None
INFO (Thread-2) [custom_components.ble_monitor.ble_parser] Unknown advertisement received for mac: xx:xx:xx:xx:xx:xxservice data: []manufacturer specific data: [b'\x0f\xff!a\x02\xb7\x00\x03\x00\x04\x1b\x94\x00\x12g\x01']local name: 8099-AFUUID16: None,UUID128: None
INFO (Thread-2) [custom_components.ble_monitor.ble_parser] Unknown advertisement received for mac: xx:xx:xx:xx:xx:xxservice data: []manufacturer specific data: [b'\x0f\xff"Q\x03\xce\x19\xaeE(\x00\x93\x05U\x00\x02']local name: 8099-AFUUID16: None,UUID128: None
INFO (Thread-2) [custom_components.ble_monitor.ble_parser] Unknown advertisement received for mac: xx:xx:xx:xx:xx:xxservice data: []manufacturer specific data: [b'\x0f\xff!a\x02\xb8\x00\x03\x00\x04\x1b\x94\x00\x12h\x01']local name: 8099-AFUUID16: None,UUID128: None
INFO (Thread-2) [custom_components.ble_monitor.ble_parser] Unknown advertisement received for mac: xx:xx:xx:xx:xx:xxservice data: []manufacturer specific data: [b'\x0f\xff"Q\x03\xc6\x19\xadE(\x00\x92\x05U\x00\x02']local name: 8099-AFUUID16: None,UUID128: None
INFO (Thread-2) [custom_components.ble_monitor.ble_parser] Unknown advertisement received for mac: xx:xx:xx:xx:xx:xxservice data: []manufacturer specific data: [b'\x0f\xff!a\x02\xb8\x00\x03\x00\x04\x1b\x94\x00\x12i\x01']local name: 8099-AFUUID16: None,UUID128: None
INFO (Thread-2) [custom_components.ble_monitor.ble_parser] Unknown advertisement received for mac: xx:xx:xx:xx:xx:xxservice data: []manufacturer specific data: [b'\x0f\xff"Q\x03\xd8\x19\xb0E(\x00\x94\x05U\x00\x02']local name: 8099-AFUUID16: None,UUID128: None
INFO (Thread-2) [custom_components.ble_monitor.ble_parser] Unknown advertisement received for mac: xx:xx:xx:xx:xx:xxservice data: []manufacturer specific data: [b'\x0f\xff!a\x02\xb9\x00\x03\x00\x04\x1b\x94\x00\x12j\x01']local name: 8099-AFUUID16: None,UUID128: None
INFO (Thread-2) [custom_components.ble_monitor.ble_parser] Unknown advertisement received for mac: xx:xx:xx:xx:xx:xxservice data: []manufacturer specific data: [b'\x0f\xff"Q\x03\xcf\x19\xadE(\x00\x93\x05U\x00\x02']local name: 8099-AFUUID16: None,UUID128: None
INFO (Thread-2) [custom_components.ble_monitor.ble_parser] Unknown advertisement received for mac: xx:xx:xx:xx:xx:xxservice data: []manufacturer specific data: [b'\x0f\xff!a\x02\xb9\x00\x03\x00\x04\x1b\x94\x00\x12k\x01']local name: 8099-AFUUID16: None,UUID128: None
INFO (Thread-2) [custom_components.ble_monitor.ble_parser] Unknown advertisement received for mac: xx:xx:xx:xx:xx:xxservice data: []manufacturer specific data: [b'\x0f\xff"Q\x03\xcd\x19\xb0E(\x00\x92\x05U\x00\x02']local name: 8099-AFUUID16: None,UUID128: None
INFO (Thread-2) [custom_components.ble_monitor.ble_parser] Unknown advertisement received for mac: xx:xx:xx:xx:xx:xxservice data: []manufacturer specific data: [b'\x0f\xff!a\x02\xb8\x00\x03\x00\x04\x1b\x94\x00\x12l\x01']local name: 8099-AFUUID16: None,UUID128: None
INFO (Thread-2) [custom_components.ble_monitor.ble_parser] Unknown advertisement received for mac: xx:xx:xx:xx:xx:xxservice data: []manufacturer specific data: [b'\x0f\xff"Q\x03\xc6\x19\xabE(\x00\x92\x05U\x00\x02']local name: 8099-AFUUID16: None,UUID128: None
INFO (Thread-2) [custom_components.ble_monitor.ble_parser] Unknown advertisement received for mac: xx:xx:xx:xx:xx:xxservice data: []manufacturer specific data: [b'\x0f\xff!a\x02\xb8\x00\x03\x00\x04\x1b\x94\x00\x12m\x01']local name: 8099-AFUUID16: None,UUID128: None
```

### PR DESCRIPTION
**Pull request by cebrusfs**
resolve #1100

I don't get much information from the posts and decide to contribute my current study result. It is right now match all values display in iOS app.

---

One question for `Pro 2` in https://github.com/custom-components/ble_monitor/blob/c995d3d5bd0b7c159e35bfc1650b012fe36d93dd/custom_components/ble_monitor/ble_parser/airmentor.py#L41-L42

I think we should align the behavior in both `Pro 2` and `S2` parsing code.

### CODE REVIEW

**Comment by Ernst79** [comment]
About the question

> I think we should align the behavior in both Pro 2 and S2 parsing code.

What should it be, according to you?

**Comment by Ernst79** [comment]
Can you also add the documentation (just copy a .md file from another airmentor sensor and modify, and add a picture (max 200x200) in the assets folder.

**Comment by cebrusfs** [comment]
> What should it be, according to you?

I am not sure if we should cover the backward competibility. If not, I slight prefer to store the data in the correct unit. 

However, if we change the fomular here, we might mix two different kind unit of data in user's db. (we store ppb as mg/m^3 in the past and now we will store mg/m^3 in mg/m^3) and might cause some analysis logic fail. Not sure what's the best practice is.

BTW, I am curious Is there any specify a particular unit and not use default unit `native_unit_of_measurement` in follow code? https://github.com/custom-components/ble_monitor/blob/e80d1f9552b9e42a6b34c1e2201398edd1c18416/custom_components/ble_monitor/const.py#L643-L652

> Can you also add the documentation (just copy a .md file from another airmentor sensor and modify, and add a picture (max 200x200) in the assets folder.

I will do so.

**Comment by cebrusfs** [comment]
Done for the doc. I think the tvoc unit issue is the last open item.

**Comment by Ernst79** [comment]
I'm not sure if the air mentor pro is the same as the air mentor pro 2 (I can't find the pro 2 on the website of air mentor). 

But for the pro, it seems that they use bbp as well, (https://www.air-mentor.com/web2017/product/air_mentor_8096ap)

We can do two things, make a second sensortype with a different unit or convert the air mentor pro 2 in the same way as you did for the S2 to ug/m3. I would propose to do the last, change the unit of the pro 2 to ug/m3. I will add a note in the release notes than.

File: custom_components/ble_monitor/ble_parser/airmentor.py [hunk 1]
```diff
@@ -11,55 +11,117 @@
...
+    if msg_type in [0x12, 0x22]:
+        # 2 unknown bytes at the end.
+        (tvoc_ppb, temp, temp_cal, humi, aqi, _, _) = unpack(">HHBBHHH", xvalue)
+
+        # Convert TVOC from ppb to mg/m^3.
+        # ref: https://www.catsensors.com/media/pdf/Sensor_Sensirion_IAM.pdf
+        M_gas = 110
+        V_m = 0.0244 * 1000
+        tvoc_ugm3 = float(tvoc_ppb) * M_gas / V_m
+        tvoc_mgm3 = tvoc_ugm3 / 1000
```
Comment (Ernst79) [inline custom_components/ble_monitor/ble_parser/airmentor.py:71]: I think tvoc should be returned in ug/m3. In `const.py` we have the unit `CONCENTRATION_MICROGRAMS_PER_CUBIC_METER`, which is ug/m3. (mg/m3 = `CONCENTRATION_MILLIGRAMS_PER_CUBIC_METER`

**Review by Ernst79 — CHANGES_REQUESTED** [review]
We also need to add a test in the `test_airmentor.py` file, if you have a raw data packet. 
If you don't have that, I can reverse engineer it for you.

File: custom_components/ble_monitor/ble_parser/airmentor.py [hunk 2]
```diff
@@ -11,55 +11,117 @@
...
+            "humidity": round(humi, 2),
+            "aqi": aqi,
+            "air quality": air_quality
+        }
+    elif msg_type in [0x11, 0x21]:
+        (co2, pm25, pm10, _, hcho_ppb, _) = unpack(">HHHHHH", xvalue)
+
+        # formaldehyde: At 25 °C, 1 ppm = 1.228 mg/m3 and 1 mg/m3 = 0.814 ppm.
+        M_hcho = 1.228
+        hcho_mgm3 = hcho_ppb * 0.001 * M_hcho
```
Comment (Ernst79) [inline custom_components/ble_monitor/ble_parser/airmentor.py:93]: Add some rounding here. I now get `'formaldehyde': 0.022104000000000002,` as a result

File: custom_components/ble_monitor/ble_parser/airmentor.py (hunk 1 above)
Comment (cebrusfs) [reply custom_components/ble_monitor/ble_parser/airmentor.py]: oops. Good catch!

File: custom_components/ble_monitor/ble_parser/airmentor.py [hunk 3]
```diff
@@ -11,55 +11,117 @@
...
+            "pm2.5": pm25,
+            "pm10": pm10,
+        }
+    else:
+        return None
+
+
+def parse_s2(msg_type, xvalue):
+    if msg_type in [0x12, 0x22]:
+        # 2 unknown bytes at the end.
```
[the remark this answers is not in the document]
Comment (cebrusfs) [reply custom_components/ble_monitor/ble_parser/airmentor.py]: Yes. it is 2 x 2 bytes. Change to 4 unknown here.

**Comment by Ernst79** [comment]
tvoc needs some rounding too

**Comment by cebrusfs** [comment]
Yep. It is a bit hard to run the pytest on Mac, so I don't run it locally. Let me use github CI/CD to run the test lol.

**Comment by Ernst79** [comment]
windows isn't much better, I do the same 😄 

Looks good to me. Nice work. Thanks!

**Review by Ernst79 — APPROVED** [review]

**Comment by cebrusfs** [comment]
Thanks for your carefully and quickly review!

**Pull request merged**